ddtrace/debugging/_function/discovery.py

```diff
@@ -135,23 +136,80 @@ def _undecorate(f, name, path):
-    # DEV: We are deliberately not handling decorators that store the original
-    #      function in __wrapped__ for now.
-    if not (f.__code__.co_name == name and abspath(f.__code__.co_filename) == path):
-        seen_functions = {f}
-        q = deque([f])  # FIFO: use popleft and append
 
-        while q:
-            next_f = q.popleft()
-
-            for g in (
-                _.cell_contents for _ in (next_f.__closure__ or []) if _isinstance(_.cell_contents, FunctionType)
-            ):
-                if g.__code__.co_name == name and abspath(g.__code__.co_filename) == path:
-                    return g
-                if g not in seen_functions:
-                    q.append(g)
-                    seen_functions.add(g)
+    def match(g):
+        return g.__code__.co_name == name and abspath(g.__code__.co_filename) == path
+
+    if _isinstance(f, FunctionType) and match(f):
+        return f
+
+    seen_functions = {f}
+    q = deque([f])  # FIFO: use popleft and append
+
+    while q:
+        g = q.popleft()
+
+        # Look for a wrapped function
+        for attr in ("__wrapped__", "func"):
+            try:
+                wrapped = object.__getattribute__(g, attr)
+                if _isinstance(wrapped, FunctionType) and wrapped not in seen_functions:
+                    if match(wrapped):
+                        return wrapped
+                    q.append(wrapped)
+                    seen_functions.add(wrapped)
+            except AttributeError:
+                pass
+
+        if _isinstance(g, partial):
+            p = cast(partial, g)
+            if match(p.func):
+                return cast(FunctionType, p.func)
+            for arg in p.args:
+                if _isinstance(arg, FunctionType) and arg not in seen_functions:
+                    if match(arg):
+                        return arg
+                    q.append(arg)
+                    seen_functions.add(arg)
+            for arg in p.keywords.values():
+                if _isinstance(arg, FunctionType) and arg not in seen_functions:
+                    if match(arg):
+                        return arg
+                    q.append(arg)
+                    seen_functions.add(arg)
+
+        # Look for a closure (function decoration)
+        if _isinstance(g, FunctionType):
+            for c in (_.cell_contents for _ in (g.__closure__ or []) if _isinstance(_.cell_contents, FunctionType)):
+                if c not in seen_functions:
+                    if match(c):
+                        return c
+                    q.append(c)
+                    seen_functions.add(c)
+
+        # Look for a function attribute (method decoration)
+        # DEV: We don't recurse over arbitrary objects. We stop at the first
+        # depth level.
+        try:
+            for v in object.__getattribute__(g, "__dict__").values():
+                if _isinstance(v, FunctionType) and v not in seen_functions and match(v):
```

Comment:
nit: return v if isUnseenAndMatch(v) or just return v if match(v) - not sure what checking seen_function is important here. it is because it is faster?